File: go/0023_merge_k_sorted_lists.go
package main

import "container/heap"

// use PriorityQueue
func mergeKLists(lists []*ListNode) *ListNode {
	pq := make(PQ, 0)
	for _, node := range lists {
		if node != nil {
			pq = append(pq, node)
		}
	}

	if len(pq) == 0 {
		return nil
	}
	heap.Init(&pq)

	head := &ListNode{}
	dummyHead := head

	for len(pq) > 0 {
		min := heap.Pop(&pq)
		minNode := min.(*ListNode)
		head.Next = minNode
		head = head.Next

		if minNode.Next != nil {
			heap.Push(&pq, minNode.Next)
		}
	}
	return dummyHead.Next
}

type PQ []*ListNode

func (pq PQ) Len() int {
	return len(pq)
}

func (pq PQ) Swap(a, b int) {
	pq[a], pq[b] = pq[b], pq[a]
}

func (pq PQ) Less(a, b int) bool {
	return pq[a].Val < pq[b].Val
}

func (pq *PQ) Push(nodeInterface interface{}) {
	node := nodeInterface.(*ListNode)
	*pq = append(*pq, node)
}

func (pq *PQ) Pop() interface{} {
	old := *pq
	lastNode := old[len(*pq)-1]
	*pq = old[:len(*pq)-1]
	return lastNode
}

// Apply merge two list alrogithm
func mergeKLists(lists []*ListNode) *ListNode {
	listsLen := len(lists)
	if listsLen == 0 {
		return nil
	} else if listsLen == 1 {
		return lists[0]
	}

	for len(lists) > 1 {
		list1, list2 := lists[0], lists[1]
		merged := mergeLists(list1, list2)

		lists = lists[2:]
		lists = append(lists, merged)
	}
	return lists[0]
}

func mergeLists(listA, listB *ListNode) *ListNode {
	res := &ListNode{}
	dummyHead := res
	for listA != nil || listB != nil {
		if listA == nil {
			res.Next = listB
			break
		}
		if listB == nil {
			res.Next = listA
			break
		}

		if listA.Val < listB.Val {
			res.Next = listA
			res = res.Next
			listA = listA.Next
		} else {
			res.Next = listB
			res = res.Next
			listB = listB.Next
		}
	}
	return dummyHead.Next
}

File: python/0023_merge_k_sorted_lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
import heapq


class MyListNode(ListNode):
    def __init__(self, val, nex):
        self.val = val
        self.next = nex

    def __lt__(self, other):
        return self.val < other.val


class Solution:
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """

        heap = [MyListNode(l.val, l.next) for l in lists if l is not None]
        heapq.heapify(heap)

        res = ListNode(-1)
        cur = res
        while heap:
            smallest = heapq.heappop(heap)
            if smallest.next is not None:
                heapq.heappush(
                    heap, MyListNode(smallest.next.val, smallest.next.next))
            cur.next = smallest
            cur = cur.next
        cur.next = None
        return res.next


class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if not lists:
            return []

        while len(lists) > 1:
            i = 0
            newlists = []
            while i + 1 < len(lists):
                l1 = lists[i]
                l2 = lists[i + 1]
                newlists.append(merge(l1, l2))
                i += 2
            if i == len(lists) - 1:
                newlists.append(lists[-1])
            lists = newlists
        return lists[0]


def merge(l1, l2):
    dummy = ListNode(0)
    cur = dummy
    while l1 or l2:
        if l1 and l2:
            if l1.val > l2.val:
                cur.next = l2
                l2 = l2.next
            else:
                cur.next = l1
                l1 = l1.next
        elif l1:
            cur.next = l1
            l1 = None
        else:
            cur.next = l2
            l2 = None
        cur = cur.next
    return dummy.next

